specs/typography/heading.md

```diff
@@ -0,0 +1,241 @@
+
+- FAST-DNA (React): [Heading](https://explore.fast.design/components/heading), [Subeading](https://explore.fast.design/components/subheading)
+- [Material Design](https://material.io/design/typography/the-type-system.html)
+- [Bootstrap](https://getbootstrap.com/docs/4.4/content/typography)
+- [Ant Design](https://ant.design/components/typography)
+- [Mineral UI](https://mineral-ui.netlify.app/typography#display-text)
+- [Semantic UI](https://semantic-ui.com/elements/header.html)
+- [Fluent UI](https://fluentsite.z22.web.core.windows.net/components/header/definition)
+
+---
+
+## Design
+
+A heading can be made from any size in the type ramp.
+
+![](./images/heading-examples.png)
+
+### API
+
+Pseudo interfaces: (*Could be functions, decorators, design system properties, or otherwise*)
+
+#### Basic heading style map
+
+Bare-bones styles applied manually.
+
+```TypeScript
+headingStyles.basicRamp()
+```
+
+Produces css classes that can be applied to any text elements. The classes are consistently named for relation to the raw type ramp.
+
+*Supports use cases 1 & 3, feature 1*
+
+```CSS
+.fast-heading-minus-2 ...
+
+.fast-heading-minus-1 ...
+
+.fast-heading-base {
+    font-size: var(--fast-type-ramp-base-size);
+    line-height: var(--fast-type-ramp-base-height);
+    font-weight: var(--fast-heading-font-weight);
+}
+
+.fast-heading-plus-1 ...
+
+...
+```
+
+Usage:
+
+```HTML
+<h1 class="fast-heading-plus-3">If all the world is a stage, I want better lighting</h1>
+
+<span class="fast-heading-minus-2">If all the world is a stage, I want better lighting</span>
+```
+
+![](./images/basic-heading-samples.png)
+
+#### Scoped standard heading element map
+
+The easiest option for accessible content when you know how many levels you have.
+
+```TypeScript
+headingStyles.standardElementMap(levels: 3, includeBase: false)
+```
+
+Applies the styles to the standard heading elements for the desired number of levels so the adaptive styling is used without needing to add and maintain class names. Starts directly above the base type ramp size or optionally includes that size depending on stylistic preference.
+
+*Supports use case 2, features 2 & 3*
+
+```CSS
+h1, h2, h3, h4, h5, h6 {
+    font-weight: var(--fast-heading-font-weight);
+}
+
+body {
+    font-size: var(--fast-type-ramp-base-size);
+    line-height: var(--fast-type-ramp-base-height);
+}
+
+/* Three levels, above base */
+
+h1 {
+    font-size: var(--fast-type-ramp-plus-3-size);
+    line-height: var(--fast-type-ramp-plus-3-height);
+}
+
+h2 {
+    font-size: var(--fast-type-ramp-plus-2-size);
+    line-height: var(--fast-type-ramp-plus-2-height);
+}
+
+h3 {
+    font-size: var(--fast-type-ramp-plus-1-size);
+    line-height: var(--fast-type-ramp-plus-1-height);
+}
+
+h4, h5, h6 { /* This rule may not be necessary, but it helps identify usage or config issues. */
+    font-size: var(--fast-type-ramp-base-size);
+    line-height: var(--fast-type-ramp-base-height);
+}
+```
```

Comment:
Is this something we would set, or is this something someone would implement based on guidance?

Reply:
Is this referring to the whole css block? If so I was intending that it's the output of the `standardElementMap` interface. Per my above comment, I thought this would most likely be configured at the base of your app, and inherited throughout the dom. Could also be done for a particular section, say with different density or a different type ramp.

Reply:
Yeah, the whole block, I think my primary question was how we're exposing this as a distributed set of code. Is this particularly guidance and naming?

Reply:
We just tweaked the naming of the underlying ramp in Nick's PR, but otherwise the intent was that logic in that function generated this block or equivalent according to the parameters. The logic may not be clear though, so for instance if you asked for 4 levels of headings we would add a definition for h4 and shift all the values (+x) by one. Similarly if you set the option to have your smallest heading be the same as your body we would shift by one the other way.

The idea is quick mapping to the standard elements you configure that you're going to use, but sized according to the design system and progressing from your body text.

This is my preferred and recommended usage pattern, so I want to make sure it's defined clearly. Maybe it needs another example.

Reply:
> The idea is quick mapping to the standard elements you configure that you're going to use, but sized according to the design system and progressing from your body text.

We had discussed having other methods of applying these (not just affixing to the principle of mapping directly to standard elements). Is that not the case with the proposal here? I think I'm still trying to understand what is guidance vs. implementation. If our utility only supports generating a font ramp for standard elements I think it may be too prescriptive. If the utility generates a ramp and ties it to either a list of values (my own classes) or semantic elements then I think that's fine. 

I like the intent of using semantic elements, but I still get the feeling that our utility here assumes that most documents flow like standard markdown or documentation sites, whereas the visual hierarchy for many marketing sites and apps varies based on context as well. The visual representation of the elements may not always be directly tied to their DOM hierarchy. If this utility could support and enable *both* of those scenarios I think it adds a lot of value.

Just to be clear, not trying to be difficult here :). While it's fine to have our default implementations be more prescriptive and default towards our preference/guidance, I think we need to be careful to not impose that guidance through implementation. Just trying to parse what the export is and whether the code enforces this principle of styling the semantic elements as it directly relates to hierarchy.

Reply:
My comment a little further down might address your concern. It at least allows a structured way for someone to flexibly define whatever style mappings they need. I still like the idea of having a basic one that we think is useful, but we could provide examples in the docs for someone to use instead as well.